File: wnode-sdk-go/mesh/persistence_fs.go
package mesh

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

type FileSystemPersistenceAdapter struct {
	dataDir string
}

func NewFileSystemPersistenceAdapter(dataDir string) *FileSystemPersistenceAdapter {
	os.MkdirAll(dataDir, 0755)
	return &FileSystemPersistenceAdapter{dataDir: dataDir}
}

func (a *FileSystemPersistenceAdapter) WriteSnapshot(name string, snapshot MeshSnapshot) error {
	path := filepath.Join(a.dataDir, name+".json")
	b, _ := json.MarshalIndent(snapshot, "", "  ")
	return ioutil.WriteFile(path, b, 0644)
}

func (a *FileSystemPersistenceAdapter) ReadSnapshot(name string) (*MeshSnapshot, error) {
	path := filepath.Join(a.dataDir, name+".json")
	b, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var s MeshSnapshot
	json.Unmarshal(b, &s)
	return &s, nil
}

func (a *FileSystemPersistenceAdapter) AppendEvent(logName string, event MeshEvent) error {
	path := filepath.Join(a.dataDir, logName+".jsonl")
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	b, _ := json.Marshal(event)
	f.Write(append(b, '\n'))
	return nil
}

func (a *FileSystemPersistenceAdapter) ReadEvents(logName string) ([]MeshEvent, error) {
	path := filepath.Join(a.dataDir, logName+".jsonl")
	b, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, err
	}
	lines := strings.Split(string(b), "\n")
	var events []MeshEvent
	for _, l := range lines {
		if strings.TrimSpace(l) == "" {
			continue
		}
		var e MeshEvent
		json.Unmarshal([]byte(l), &e)
		events = append(events, e)
	}
	return events, nil
}

File: packages/wnode-sdk-ts/src/mesh/persistence_fs.ts
import * as fs from 'fs';
import * as path from 'path';
import { MeshSnapshot } from './snapshot';
import { MeshEvent } from './journal';

export class FileSystemPersistenceAdapter {
  constructor(private readonly dataDir: string) {
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }
  }

  public writeSnapshot(snapshotName: string, snapshot: MeshSnapshot): void {
    const snapshotPath = path.join(this.dataDir, `${snapshotName}.json`);
    const data = JSON.stringify(snapshot, null, 2);
    // Write atomically
    const tempPath = `${snapshotPath}.tmp`;
    fs.writeFileSync(tempPath, data, 'utf8');
    fs.renameSync(tempPath, snapshotPath);
  }

  public readSnapshot(snapshotName: string): MeshSnapshot | null {
    const snapshotPath = path.join(this.dataDir, `${snapshotName}.json`);
    if (!fs.existsSync(snapshotPath)) {
      return null;
    }
    const data = fs.readFileSync(snapshotPath, 'utf8');
    return JSON.parse(data) as MeshSnapshot;
  }

  public appendEvent(logName: string, event: MeshEvent): void {
    const logPath = path.join(this.dataDir, `${logName}.jsonl`);
    const line = JSON.stringify(event) + '\n';
    fs.appendFileSync(logPath, line, 'utf8');
  }

  public readEvents(logName: string): MeshEvent[] {
    const logPath = path.join(this.dataDir, `${logName}.jsonl`);
    if (!fs.existsSync(logPath)) {
      return [];
    }
    const data = fs.readFileSync(logPath, 'utf8');
    const lines = data.split('\n').filter(l => l.trim() !== '');
    return lines.map(l => JSON.parse(l) as MeshEvent);
  }
}
